forge/src/mixins/java/org/spongepowered/forge/mixin/core/server/network/ServerLoginPacketListenerImplMixin_Forge.java

```diff
@@ -24,51 +24,27 @@
  */
 package org.spongepowered.forge.mixin.core.server.network;
 
-import com.mojang.authlib.GameProfile;
 import net.minecraft.network.protocol.login.ServerLoginPacketListener;
-import net.minecraft.network.protocol.login.ServerboundHelloPacket;
 import net.minecraft.server.network.ServerLoginPacketListenerImpl;
-import org.objectweb.asm.Opcodes;
 import org.spongepowered.asm.mixin.Mixin;
 import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
-import org.spongepowered.common.bridge.server.network.ServerLoginPacketListenerImplBridge;
-
-import javax.annotation.Nullable;
+import org.spongepowered.common.bridge.network.ServerLoginPacketListenerImplBridge;
 
 @Mixin(ServerLoginPacketListenerImpl.class)
 public abstract class ServerLoginPacketListenerImplMixin_Forge implements ServerLoginPacketListener, ServerLoginPacketListenerImplBridge {
 
     // @formatter:off
-    @Shadow @Nullable public GameProfile gameProfile;
-    @Shadow protected abstract GameProfile shadow$createFakeProfile(GameProfile $$0);
-
-    //@formatter:on
-
-    /**
-     * @reason Forge alters {@link ServerLoginPacketListenerImpl#handleHello(ServerboundHelloPacket)} to switch
-     * from {@link ServerLoginPacketListenerImpl.State#READY_TO_ACCEPT} to {@link ServerLoginPacketListenerImpl.State#NEGOTIATING}
-     * to verify mods. We need to clarify that the player is authenticating finally at the point when we really
-     * transition to READY_TO_ACCEPT.
-     *
-     * @param ci The callback info to cancel
-     */
-    @Inject(method = "tick()V",
-        at = @At(
-            value = "FIELD",
-            target = "Lnet/minecraft/server/network/ServerLoginPacketListenerImpl$State;READY_TO_ACCEPT:Lnet/minecraft/server/network/ServerLoginPacketListenerImpl$State;",
-            opcode = Opcodes.GETSTATIC),
-        cancellable = true)
-    private void forge$fireAuthEventOffline(final CallbackInfo ci) {
-        // Move this check up here, so that the UUID isn't null when we fire the event
-        if (!this.gameProfile.isComplete()) {
-            this.gameProfile = this.shadow$createFakeProfile(this.gameProfile);
-        }
+    @Shadow private ServerLoginPacketListenerImpl.State state;
+    // @formatter:on
 
-        if (this.bridge$fireAuthEvent()) {
-            ci.cancel();
+    @Inject(method = "tick", at = @At("HEAD"))
+    private void impl$onTick(final CallbackInfo ci) {
+        // In SpongeVanilla we do channel registration during this state, not sure if we need to do anything in SpongeForge
+        if (this.state == ServerLoginPacketListenerImpl.State.NEGOTIATING) {
```

Comment:
Why does this change exist?

Reply:
See https://github.com/SpongePowered/Sponge/blob/api11/forge/src/mixins/java/org/spongepowered/common/mixin/core/server/network/ServerLoginPacketListenerImplMixin.java#L146
and https://github.com/SpongePowered/Sponge/blob/api11/forge/vanilla/src/mixins/java/org/spongepowered/vanilla/mixin/core/server/network/ServerLoginPacketListenerImplMixin_Vanilla.java#L80

After auth, we switch to state `NEGOTIATING` which is unused by vanilla MC. We use this state in SpongeVanilla to register channels and then we go to `VERIFYING` which will be handled by vanilla MC.

I had to replicate the `NEGOTIATING` -> `VERIFYING` switch in SpongeForge otherwise the login stay forever in state `NEGOTIATING`.